### PR TITLE
Various fixes, multi threaded and support for HA using DDB

### PR DESCRIPTION
**Pull request by dpgaspar**
Hi,

I know these are a lot of changes. But this plugin is under some moderate load on a testing environment and is now working great. 

The DynamoDB option, is for HA, this can be accomplished also by using EFS. Yet it's probably a better option for multi-threads

### CODE REVIEW

**Comment by nunofernandes** [comment]
As you are using DynamoDB, can you please update Readme.md to have the correct policy for using DynamoDB (update table state_ddb_table).

**Comment by nunofernandes** [comment]
+1 Please merge this!

**Comment by dpgaspar** [comment]
Hi,

First of all the DDB option is optional.
I do think it brings plugin stability and speed. Of course there are other options also, like still keeping the state on files that live on an NFS. DDB is faster and lowers the probability of a race condition (thought each thread consumes it's own stream and each stream is kept on a different file).

I did not understand what is the approach on HA made by out_elasticsearch_dynamic, could you detail that?

**Comment by cosmo0920** [comment]
> I did not understand what is the approach on HA made by out_elasticsearch_dynamic, could you detail that?

My suggestion is out_cloudwatch_logs_ha.rb(CloudwatchLogsHAOutput) or something inherits CloudwatchLogsOutput class and implements HA mechanism.
This approach is similar to out_elasticsearch_dynamic.

**Comment by balan2010** [comment]
+1 any idea when this would be merged

File: lib/fluent/plugin/in_cloudwatch_logs.rb [hunk 1]
```diff
@@ -87,33 +100,128 @@ def state_file_for(log_stream_name)
...
+        emit(log_stream_name, event)
+        c = c + 1
+      end
+      log.info("Cloudwatch #{@log_group_name} Emited #{c} events")
+    end
+
+    def consumer()
+      threads = []
+      @thread_num.times do
+        threads << Thread.new do
```
Comment (cosmo0920) [inline lib/fluent/plugin/in_cloudwatch_logs.rb]: Please use `:thread` plugin helper if replaceable.

**Comment by kladiv** [comment]
+1 to merge the fix for cloudwatch throttling error

**Pull request closed without merging**